rust/src/cli/persist_nic.rs

```diff
@@ -195,49 +116,146 @@ pub(crate) fn run_persist_immediately(
-        if !name.starts_with(PERSIST_FILE_PREFIX) {
-            continue;
+        if let Some(iface_name) = extract_iface_names_from_link_file(&file_name)
+        {
+            log::info!("Found persisted NIC({iface_name}) file: {file_name}");
+            pinned_ifaces
+                .insert(iface_name.to_string(), netdir.join(file_name));
         }
-        log::info!("Found persisted NIC file: {name}");
-        n += 1;
     }
-    if n == 0 {
+
+    if pinned_ifaces.is_empty() {
         log::info!("No persisted NICs found");
+        return Ok("".to_string());
     }
 
-    let state = gather_state()?;
-    process_interfaces(&state, |iface, mac| {
-        let iface_name = iface.name();
-        log::info!(
-            "NOTE: would persist the interface with MAC {mac} to interface name {iface_name}"
-        );
-        Ok(())
+    for (iface_name, file_path) in pinned_ifaces.drain() {
+        if !is_nmstate_generated_systemd_link_file(&file_path) {
+            log::info!(
+                "File {} is not generated by nmstate, ignoring",
+                file_path.display()
+            );
+            continue;
+        }
+        let systemd_iface_name =
+            match get_systemd_preferred_iface_name(&iface_name) {
+                Ok(i) => i,
+                Err(e) => {
+                    log::error!(
+                        "Failed to retrieve systemd preferred \
+                        iface name for {iface_name}: {e}"
+                    );
+                    continue;
+                }
+            };
+        if systemd_iface_name == iface_name {
+            if dry_run {
+                log::info!(
+                    "The generated {} file has no effect for \
+                    interface {iface_name}, will remove if not \
+                    in dry-run mode",
+                    file_path.display()
+                );
+            } else {
+                log::info!(
+                    "The generated {} file has no effect for \
+                    interface {iface_name}, removing",
+                    file_path.display()
+                );
+                std::fs::remove_file(file_path)?;
+            }
+        } else {
+            log::info!(
+                "Systemd generate interface name \
+                '{systemd_iface_name}' != pinned name '{iface_name}', \
+                will keep config file {}",
+                file_path.display()
+            );
+        }
+    }
+    Ok("".to_string())
+}
+
+// With `NamePolicy=keep kernel database onboard slot path` in systemd configure
+// in RHEL 8 and 9. Assuming `keep, kernel and database` all return NULL,
+// Systemd will use interface name in the order of:
+//  * `ID_NET_NAME_ONBOARD`
+//  * `ID_NET_NAME_SLOT`
+//  * `ID_NET_NAME_PATH`
+pub(crate) fn get_systemd_preferred_iface_name(
+    iface_name: &str,
+) -> Result<String, CliError> {
+    let output: String = String::from_utf8(
+        std::process::Command::new("udevadm")
```

Comment:
And if we have a root, use `chroot`.

Reply:
This `chroot /new_root udevadm` require `sysfs` been mounted in new root, is that true for openshift use case?

Reply:
Yep, we are (recursively) bind mounting the host `/`, so we'll see everything the host does.